lisa/energy_model.py

```diff
@@ -1034,13 +1035,43 @@ def read_sge_file(path):
         def read_active_states(cpu, domain_level):
             cap_states_path = sge_path(cpu, domain_level, 0, 'cap_states')
             cap_states_strs = read_sge_file(cap_states_path).split()
+            nr_cap_states_path = sge_path(cpu, domain_level, 0, 'nr_cap_states')
+            nr_cap_states_strs = read_sge_file(nr_cap_states_path).split()
+            # there are potentially two formats for this data which can be
+            # differentiated by knowing how many strings were obtained when
+            # we split cap_states *and* how many cap states there are.
+            # If the split has 2x the number of states, the reported states
+            # are from a kernel without frequency-model support and there
+            # two values per state. If the split has 3x the number of states
+            # then the reported states are from a kernel which *has*
+            # frequency model support, and each state has three values to parse.
+            nr_values = len(cap_states_strs)
+            nr_states  = int(nr_cap_states_strs[0])
+            if (nr_states * 2) == nr_values:
+                freq_model_supported = False
+            elif (nr_states * 3) == nr_values:
+                freq_model_supported = True
+            else:
+                raise TargetStableError('Unsupported cap_states format '
+                                  'cpu={} domain_level={} path={}'.format(cpu, domain_level, cap_states_path))
 
+            # In a two-value system (no frequency model),
             # cap_states lists the capacity of each state followed by its power,
             # in increasing order. The `zip` call does this:
             #   [c0, p0, c1, p1, c2, p2] -> [(c0, p0), (c1, p1), (c2, p2)]
-            cap_states = [ActiveState(capacity=int(c), power=int(p))
+            if (not freq_model_supported):
+                cap_states = [ActiveState(capacity=int(c), power=int(p))
```

Comment:
@douglas-raillard-arm mentioned in #657 against master that we should probably update the grouping. He wrote:
iterating N items at a time seems to be a common pattern, it may be worth adding the "grouper" from itertools recipes to a helper module (or just locally here with a #TODO to move it, to ease forward porting to LISA-next):
https://docs.python.org/3/library/itertools.html#itertools-recipes

```
from itertools import zip_longest
def grouper(iterable, n, fillvalue=None):
    "Collect data into fixed-length chunks or blocks"
    # grouper('ABCDEFG', 3, 'x') --> ABC DEF Gxx"
    # Since the same iterator is used, it will yield a new item every time zip call next() on it
    args = [iter(iterable)] * n
    return zip_longest(*args, fillvalue=fillvalue)
```

_Originally posted by @douglas-raillard-arm in https://github.com/_render_node/MDIzOlB1bGxSZXF1ZXN0UmV2aWV3VGhyZWFkMTQxODU0MzMzOnYy/pull_request_review_threads/discussion_